api/src/main/java/io/opencensus/trace/propagation/B3Format.java

```diff
@@ -0,0 +1,143 @@
+ * void onSendRequest() {
+ *   try (Scope ss = tracer.spanBuilder("Sent.MyRequest").startScopedSpan()) {
+ *     Map<HeaderName, String> b3Headers  =
+ *         b3Format.toHeaders(tracer.getCurrentContext().context());
+ *     // Send the request including the b3Headers and wait for the response.
+ *   }
+ * }
+ * }</pre>
+ *
+ * <p>Example of usage on the server:
+ *
+ * <pre>{@code
+ * private static final Tracer tracer = Tracing.getTracer();
+ * private static final B3Format b3Format = Tracing.getPropagationComponent().getB3Format();
+ * void onRequestReceived() {
+ *   // Get the B3 headers from the request.
+ *   Map<HeaderName, String> b3Headers = ...
+ *   SpanContext spanContext = b3Format.fromHeaders(b3Headers);
+ *   try (Scope ss =
+ *            tracer.spanBuilderWithRemoteParent("Recv.MyRequest", spanContext).startScopedSpan()) {
+ *     // Handle request and send response back.
+ *   }
+ * }
+ * }</pre>
+ */
+public abstract class B3Format {
+  public enum HeaderName {
+    X_B3_TRACE_ID("X─B3─TraceId"),
+    X_B3_SPAN_ID("X─B3─SpanId"),
+    X_B3_PARENT_SPAN_ID("X─B3─ParentSpanId"),
+    X_B3_SAMPLED("X─B3─Sampled"),
+    X_B3_FLAGS("X-B3-Flags");
+
+    private final String value;
+    private final String lowerValue;
+
+    private HeaderName(String value) {
+      this.value = value;
+      this.lowerValue = value.toLowerCase(Locale.US);
+    }
+
+    /**
+     * Returns the value of this header used in HTTP requests.
+     *
+     * @return the value of this header used in HTTP requests.
+     */
+    public String getValue() {
+      return value;
+    }
+
+    /**
+     * Returns the ASCII value of this header (lowercase).
+     *
+     * @return the ASCII value of this header (lowercase).
+     */
+    public String getLowerValue() {
+      return lowerValue;
+    }
+  }
+
+  private static final NoopB3Format NOOP_B3_FORMAT = new NoopB3Format();
+
+  /**
+   * Serializes a {@link SpanContext} using the b3-propagation format.
+   *
+   * @param spanContext the {@code SpanContext} to serialize.
+   * @return a set of headers that must be propagated on the wire.
+   * @throws NullPointerException if the {@code spanContext} is {@code null}.
+   */
+  public abstract Map<HeaderName, String> toHeaders(SpanContext spanContext);
+
+  /**
+   * Parses the {@link SpanContext} from the b3-propagation format.
+   *
+   * @param headers a set of headers from which the {@code SpanContext} will be parsed.
+   * @return the parsed {@code SpanContext}.
+   * @throws NullPointerException if the {@code headers} is {@code null}.
+   */
+  public abstract SpanContext fromHeaders(Map<HeaderName, String> headers) throws ParseException;
+
+  /**
+   * Returns the no-op implementation of the {@code B3Format}.
+   *
+   * @return the no-op implementation of the {@code B3Format}.
+   */
+  static B3Format getNoopB3Format() {
+    return NOOP_B3_FORMAT;
+  }
+
+  private static final class NoopB3Format extends B3Format {
```

Comment:
food for thought, but this is what brave uses. In practice `C` is usually an httprequest object. We have a supporting type Getter and Setter to replace elements in the carrier. This prevents us from having to play any games with making a copy of a map especially as almost no http api has request headers as a string-> string map.

```
 /**
   * Used to send the trace context downstream. For example, as http headers.
   */
  public interface Injector<C> {
    /**
     * Usually calls a setter for each propagation field to send downstream.
     *
     * @param traceContext possibly unsampled.
     * @param carrier holds propagation fields. For example, an outgoing message or http request.
     */
    void inject(TraceContext traceContext, C carrier);
  }

  /**
   * Used to join an incoming trace. For example, by reading http headers.
   *
   * @see brave.Tracer#joinSpan
   */
  public interface Extractor<C> {

    /**
     * Returns either a trace context or sampling flags parsed from the carrier. If nothing was
     * parsable, sampling flags will be set to {@link SamplingFlags#EMPTY}.
     *
     * @param carrier holds propagation fields. For example, an incoming message or http request.
     */
    TraceContextOrSamplingFlags extract(C carrier);
  }
```

Reply:
Does Brave implement Injector and Extractor, and then the user implements a Getter and Setter specific to their request type?

Reply:
For simple string things like http headers, it is easy.

```java

or more simply for http:

```java
  static final Propagation.Setter<MultivaluedMap, String> SETTER = MultivaluedMap::putSingle;

  @Inject TracingClientFilter(HttpTracing httpTracing) {
--snip--
    injector = httpTracing.tracing().propagation().injector(SETTER);
  }

--snip--
    injector.inject(span.context(), request.getHeaders());
```

gRPC uses predefined keys, so requires a bit more coordination
```java
enum AsciiMetadataKeyFactory implements Propagation.KeyFactory<Key<String>> {
  INSTANCE;

  @Override public Key<String> create(String name) {
    return Key.of(name, Metadata.ASCII_STRING_MARSHALLER);
  }
}
...
  static final Setter<Metadata, Metadata.Key<String>> SETTER = (metadata, key, value) -> {
    metadata.removeAll(key);
    if (value != null) metadata.put(key, value);
  };
...
  final TraceContext.Injector<Metadata> injector;

  TracingClientInterceptor(Tracing tracing) {
    tracer = tracing.tracer();
    injector = tracing.propagationFactory().create(AsciiMetadataKeyFactory.INSTANCE).injector(SETTER);
  }
...
    @Override
    public void start(Listener<RespT> responseListener, Metadata headers) {
      injector.inject(span.context(), headers);
```